aiomisc/aggregate.py

```diff
@@ -0,0 +1,144 @@
+from typing import Any, Awaitable, Callable, Iterable, List, Optional
+
+AggFunc = Callable[[Any], Awaitable[Iterable]]
+
+
+class Aggregator:
+
+    def __init__(
+            self, func: AggFunc, *,
+            leeway_ms: float, max_count: int = None,
+    ):
+        has_variadic_positional = any((
+            parameter.kind == Parameter.VAR_POSITIONAL
+            for parameter in inspect.signature(func).parameters.values()
+        ))
+        if not has_variadic_positional:
+            raise ValueError(
+                "Function must accept variadic positional arguments",
+            )
+
+        if max_count is not None and max_count <= 0:
+            raise ValueError("max_count must be positive int or None")
+
+        if leeway_ms <= 0:
+            raise ValueError("leeway_ms must be positive float")
+
+        self._func = func
+        self._max_count = max_count
+        self._leeway = leeway_ms / 1000
+        self._t = None          # type: Optional[float]
+        self._args = []         # type: list
+        self._futures = []      # type: List[Future]
+        self._event = Event()   # type: Event
+        self._lock = Lock()     # type: Lock
+
+    @property
+    def max_count(self) -> Optional[int]:
+        return self._max_count
+
+    @property
+    def leeway_ms(self) -> float:
+        return self._leeway * 1000
+
+    @property
+    def count(self) -> int:
+        return len(self._args)
+
+    def _clear(self) -> None:
+        self._t = None
+        self._args = []
+        self._futures = []
+        self._event = Event()
+        self._lock = Lock()
+
+    async def _execute(self, *, args: list, futures: List[Future]) -> None:
+        try:
+            results = await self._func(*args)
+        except CancelledError:
+            # Other waiting tasks can try to finish the job instead.
+            raise
+        except Exception as e:
+            self._set_exception(e, futures)
+            return
+
+        self._set_results(results, futures)
+
+    def _set_results(self, results: Iterable, futures: List[Future]) -> None:
+        for future, result in zip(futures, results):
+            if not future.done():
+                future.set_result(result)
+
+    def _set_exception(
+            self, exc: Exception, futures: List[Future],
+    ) -> None:
+        for future in futures:
+            if not future.done():
+                future.set_exception(exc)
+
+    async def aggregate(self, arg: Any) -> Any:
+        if not self._t:
+            self._t = monotonic()
+        t = self._t
+
+        args = self._args           # type: List
+        futures = self._futures     # type: List
+        event = self._event         # type: Event
+        lock = self._lock           # type: Lock
+        args.append(arg)
+        future = Future()           # type: Future
+        futures.append(future)
+
+        if self.count == self.max_count:
+            event.set()
+            self._clear()
```

Comment:
Ниже не почистится?

Reply:
Need to clear it as soon as possible. Otherwise, another call may happen during `await wait_for` and it will not start a new aggregation batch, but will reuse this one.